### PR TITLE
feat(rdb_load): add support for loading huge sets

### PR DESCRIPTION
**Pull request by andydunstall**
Adds support for loading huge sets (https://github.com/dragonflydb/dragonfly/issues/3760)

### CODE REVIEW

File: src/server/rdb_load.cc [hunk 1]
```diff
@@ -1487,25 +1508,47 @@ auto RdbLoaderBase::ReadLzf() -> io::Result<LzfString> {
...
+  }
 
   if (rdbtype == RDB_TYPE_SET_WITH_EXPIRY) {
     len *= 2;
   }
 
+  /*
+   * TODO
+   *
+   * I'm quite confused by the existing logic here, were we limit
```
Comment (romange) [inline src/server/rdb_load.cc:1524]: If i remember correctly, it was more efficient to allocate smaller arrays than a single huge array.
And I agree with you - lets start with a single "static" rule of upto kMaxBlobLen for streaming.

File: src/server/rdb_load.cc [hunk 2]
```diff
@@ -2556,63 +2619,83 @@ void RdbLoader::ResizeDb(size_t key_num, size_t expire_num) {
...
+    if (ec) {
+      VLOG(1) << "ReadObj error " << ec << " for key " << item->key;
+      return ec;
+    }
 
-  if (!load_unowned_slots_ && cluster::IsClusterEnabled()) {
-    const cluster::ClusterConfig* cluster_config = cluster::ClusterFamily::cluster_config();
-    if (cluster_config != nullptr && !cluster_config->IsMySlot(item->key)) {
-      return kOk;  // Ignoring item
+    if (pending_read_.remaining > 0) {
```
Comment (romange) [inline src/server/rdb_load.cc:2650]: i would use `item->append` condition here and then would group the 3 ifs here, at line 2657 and 2673 accordingly.

**Review by romange — COMMENTED** [review]
looks good :)

File: src/server/rdb_load.cc (hunk 1 above)
Comment (andydunstall) [reply src/server/rdb_load.cc]: ok sure

File: src/server/rdb_load.cc [hunk 3]
```diff
@@ -2556,63 +2619,83 @@ void RdbLoader::ResizeDb(size_t key_num, size_t expire_num) {
...
-  out_buf.emplace_back(item);
-  std::move(cleanup).Cancel();
+    auto& out_buf = shard_buf_[sid];
 
-  constexpr size_t kBufSize = 128;
-  if (out_buf.size() >= kBufSize) {
-    // Despite being async, this function can block if the shard queue is full.
-    FlushShardAsync(sid);
-  }
+    out_buf.emplace_back(item);
```
[the remark this answers is not in the document]
Comment (andydunstall) [reply src/server/rdb_load.cc]: thanks updated

File: src/server/rdb_load.cc (hunk 2 above)
Comment (andydunstall) [reply src/server/rdb_load.cc:2639]: this is checking whether this is the last read in the object (which we can't infer from `append` which says whether it's the first read in the object)

> group the 3 ifs here, at line 2657 and 2673 accordingly.

i got that logic wrong (we need to check whether to discard on each read) - i've refactored

**Comment by romange** [comment]
you can use `rdb_test.cc` to add the unit test

**Comment by andydunstall** [comment]
Sure I've added a unit test

Testing manually a 5GB set with 5m entries (`debug populate 1 test 1000 rand type set elements 5000000`)

The RSS is still 10GB (5GB used), as it still only does a single flush (since kMaxBufSize x kMaxBlobLen > 8m)

It's also now ~20% slower which seems to be due to not reserving the full set up front anymore. Though can include the full set size in `Item` to reserve on first load if you think it's worth doing

Repeating the test where we flush if `pending_read.remaining > 0`, RSS < 7GB (5GB used) and it's 20% faster. Not sure if you think thats worth adding or if theres another flushing approach we should use?

**Comment by romange** [comment]
I think it's worth reserving capacity if it's easy to add. Also we can reduce now `kMaxBlobLen` to smaller number since we know large sets will just occupy all the items in `ItemsBuf`. 
using `pending_read.remaining` as a flush signal is also possible and could be a good short term heuristic.

**Comment by andydunstall** [comment]
Ok sure, I've reduced kMaxBlobLen to 4096 meaning the 5m set is split into ~10 flushes (128x4096 ~= 500k) 

And reserved the full set up front

With that loading the 5GB set takes 2.8s (compared to 3.8s before) with RSS ~7GiB (~10GiB before)

File: src/server/rdb_load.cc [hunk 4]
```diff
@@ -2505,6 +2557,11 @@ void RdbLoader::LoadItemsBuffer(DbIndex db_ind, const ItemsBuf& ib) {
       stop_early_ = true;
       break;
     }
+    if (item->load_config.append) {
```
Comment (romange) [inline src/server/rdb_load.cc:2560]: why did you add this code?

File: src/server/rdb_load.cc [hunk 5]
```diff
@@ -2549,29 +2606,79 @@ void RdbLoader::ResizeDb(size_t key_num, size_t expire_num) {
...
+
+    ShardId sid = Shard(item->key, shard_set->size());
+    item->expire_ms = settings->expiretime;
+
+    auto& out_buf = shard_buf_[sid];
+
+    out_buf.emplace_back(std::move(item));
+    std::move(cleanup).Cancel();
 
+    constexpr size_t kBufSize = 128;
```
Comment (romange) [inline src/server/rdb_load.cc:2667]: lets reduce it to 64.

File: src/server/rdb_load.h [hunk 6]
```diff
@@ -112,13 +112,40 @@ class RdbLoaderBase {
...
+  };
+
   class OpaqueObjLoader;
 
   io::Result<uint8_t> FetchType();
 
   template <typename T> io::Result<T> FetchInt();
 
   static std::error_code FromOpaque(const OpaqueObj& opaque, CompactObj* pv);
+  static std::error_code FromOpaque(const OpaqueObj& opaque, CompactObj* pv, LoadConfig config);
```
Comment (romange) [inline src/server/rdb_load.h:148]: nit: `FromOpaque(const OpaqueObj& opaque, const LoadConfig& config, CompactObj* pv);`

File: src/server/rdb_load.h [hunk 7]
```diff
@@ -261,6 +291,8 @@ class RdbLoader : protected RdbLoaderBase {
   struct ObjSettings;
 
   std::error_code LoadKeyValPair(int type, ObjSettings* settings);
+  // Returns whether to discard the read key pair.
+  bool DiscardKey(std::string_view key, ObjSettings* settings);
```
Comment (romange) [inline src/server/rdb_load.h:295]: nit: `bool ShouldDiscardKey(std::string_view key, const ObjSettings& settings) const;`

**Review by romange — COMMENTED** [review]
lgtm, plus minor comments

File: src/server/rdb_load.cc (hunk 4 above)
Comment (andydunstall) [reply src/server/rdb_load.cc:2560]: A bit of a guess given it already does `ts->TryStash` below in the non-append case

File: src/server/rdb_load.cc (hunk 5 above)
Comment (andydunstall) [reply src/server/rdb_load.cc]: sure updated

File: src/server/rdb_load.h (hunk 6 above)
Comment (andydunstall) [reply src/server/rdb_load.h]: sure updated

File: src/server/rdb_load.h (hunk 7 above)
Comment (andydunstall) [reply src/server/rdb_load.h]: sure updated

File: src/server/rdb_load.cc (hunk 4 above)
Comment (romange) [reply src/server/rdb_load.cc:2560]: Lets remove this code.
1. Data tiering does not work with to non-strings.
2. We do not want to stash values that are not yet finalized.

Comment (andydunstall) [reply src/server/rdb_load.cc:2560]: sure removed

**Review by romange — APPROVED** [review]

**Comment by romange** [comment]
Good work @andydunstall . If you want you can follow up with hash-map + zset PR as well.

**Comment by andydunstall** [comment]
> If you want you can follow up with hash-map + zset PR as well

Thanks will do
